### ISSUE TITLE
Allow handling of removing layer event before triggering removed event for napari layers

### ISSUE THREAD

**Issue by VasudhaJha** [napari/napari#5625]
## 🚀 Feature

I would like napari to synchronously run the `removing` layer event handler, followed by the `removed` layer event handler, unlike the current behavior where the latter does not wait for the former to finish. If a user decides to cancel the removal of the layer, then there should be a mechanism in the `removing` layer event handler which prevents the layer from being removed.

## Motivation

I am working on a brightest path image tracing plugin and I want to be able to provide the user with:

1. an option to save the tracings that they've generated thus far for the layer that they're planning to remove
2. an option to cancel the layer removal

The right place to handle these seems to be in handler for the `removing` event which provides information about the layer that is about to be removed. Immediately after this event, the `removed` event is triggered and the layer is removed from the napari viewer. 

I added the code to save the tracings in the `removing` event handler but the layer is removed and the `removed` event is triggered even before the handler for `removing` has finished. 

## Pitch

1. I would like the napari viewer to wait for the `removing` event handler to finish before removing the layer and triggering the `removed` event.
2. There needs to be a mechanism to prevent the layer from being removed depending on the behavior in the `removing` event handler.

cc @cudmore

**Comment by jni** [comment]
Ah! That seems like a bug. 😅 Certainly I agree with your assessment that `removing` should have an opportunity to actually do a thing. 😅 And this is all the more important since we don't have an undo.

I don't have a strong sense of what it would take to fix this, but we'd definitely welcome a PR to fix and would be happy to help get you going. I'm probably *not* the best person here but if you want you can set up a pair programming session with me at https://meet.jni.codes and we can try to muddle through it together! 😅

**Comment by akuten1298** [comment]
Hi @jni , I'm looking into this issue currently and will start working on a fix. Can you please assign me to this bug incase it's required? just wanted to make sure that there are no duplicate efforts. Thank you.

**Comment by akuten1298** [comment]
`I added the code to save the tracings in the removing event handler but the layer is removed and the removed event is triggered even before the handler for removing has finished.`

Hi @VasudhaJha , just to clarify, in here do you specifically mean that you would ideally want to your listener functionality to complete before triggering the removed event? Can you also point me to how you are specifically using this if possible? Thank you

cc: @aganders3

**Comment by akuten1298** [comment]
Hi @VasudhaJha , for the first part where you mention that removed event is triggered before the callback/listeners for removing event is triggered, I can see that all the callbacks being completed before the removed event is triggered. 

In the screenshots attached below, I've added three callbacks to the "removing" event emitter and tracked the start of both the events. When I execute this by triggering the delete layer functionality, I can see all the callbacks being executed before the "removed" event is executed. 

Please let us know if you still face an issue with this.

<img width="1222" alt="Screenshot 2023-05-24 at 11 30 34 AM" src="https://github.com/napari/napari/assets/51192400/162d7183-2adf-4c03-9f72-0d8296f16e49">

<img width="937" alt="Screenshot 2023-05-24 at 11 30 20 AM" src="https://github.com/napari/napari/assets/51192400/5f7a8b8d-65b2-498d-8786-bef68f654333">

cc: @aganders3

**Comment by VasudhaJha** [comment]
Hi @akuten1298, thanks for working on a fix for this. For me the handler of the "removing" event does get triggered before that of the "removed" event. However, the "removing" event handler's execution isn't finished before the "removed" event handler gets invoked. So if you could try adding a wait of around 5s in your "removing" event handlers, then we can confirm if it's the same case for you.

**Comment by akuten1298** [comment]
Hi @VasudhaJha, thanks for the clarification. I added 3 handlers with different sleep times for the "removing" event and 1 handler for the "removed" event without a sleep time. I could see them get execute sequentially though i.e all the handlers for the "removing" event are completed before the "removed" event/handler starts. I have added the callback and the order of execution in the terminal as screenshots below.

![Screenshot 2023-05-25 at 10 30 29 AM](https://github.com/napari/napari/assets/51192400/7373cfdc-d8b7-4b88-af67-1c6a84e3236f)

**Comment by akuten1298** [comment]
Can you also point me to your handler functionality incase it's public?

**Comment by akuten1298** [comment]
Hi @VasudhaJha I've raised a new PR comprising the required changes to facilitate stopping the deletion of a layer through a flag that can be set in callbacks/handlers.

https://github.com/napari/napari/pull/5891

**Comment by Czaki** [comment]
Hm. 
From one side I see that it is nice to have the option to prevent removing a layer. 

But from another hand, such an option may really confuse users. There are two cases:

1) User removes the layer using the button. Then we should at least inform the user that removing the layer was blocked. In an ideal situation, this should contain information on which plugin prevents this. Otherwise, I expect the Issue "Remove button not working" 
2) Layer is deleted using API. In such a situation, silent prevention of removing layers is a bug. It is a valid assumption that after calling `del viewer.layers[name]` the layer is removed. So this feature should not be enabled when using API or it should raise an exception. 

Maybe the better solution will be to add `protected` attribute to the layer? And if the layer is marked as protected then show a dialog with the question on the side of napari? 

cc: @jni

**Comment by aganders3** [comment]
I don't love the idea of a new attribute on the layer, but I like the dialog. Perhaps instead this dialog could appear along with the mechanism proposed in #5891. That is - if a callback tries to cancel the deletion, a dialog is shown. I think @akuten1298 had some kind of prototype of this already. I don't have a good idea for indicating which plugin/callback is (attempting) blocking deletion though.

I agree with your thoughts on deletion via API vs. GUI - definitely should be no silent prevention of removal. I feel proper documentation for devs here to indicate that the feature will _only_ block deletion via GUI would be enough and prefer that to raising an exception. Otherwise I guess plugin devs would need to inspect the event source to avoid raising?

**Comment by jni** [comment]
@Czaki maybe that comment (about deleting layers) and the resulting discussion should be moved to #5891? I feel like this particular issue is more general. For example, I might want to have a plugin where *if* the plugin created the layer *and* the layer has metadata about which filename it belongs to, then when the layer is being deleted I want to synchronise the latest changes before deleting the layer. What do you think?

**Comment by Czaki** [comment]
@jni there is a removing event of the layer list that is emitted before the layer is removed. And using this event you could implement this synchronisation.

**Comment by jni** [comment]
yes but I thought (I might have missed a resolution somewhere) that the `removing` event didn't wait for its callbacks to complete before continuing with the removal?

**Comment by Czaki** [comment]
Napari events at this moment are 100% synchronous. So it will wait on callbacks until the callback is not propagated to another thread it will wait. And even in such a situation, it is enough to store a reference to layer to prevent the garbage collector from removing the data. 

Or one could simply connect to the `removed` event that stores the reference to the layer and use it to save data on the drive.

**Comment by Czaki** [comment]
hm. But the option to connect to the event of removing a given layer, not any layer sounds interesting.
